File: AEDS2/estruturas_Abstratas_TAD/fila/filaCabeca.c
#include <stdio.h>
#include <stdlib.h>

typedef struct No {
    int dado;
    struct No *prox;
} No;

typedef struct {
    No *cabeca; // nó cabeça
    No *fim;
} Fila;

void inicializar(Fila *F) {
    F->cabeca = (No*) malloc(sizeof(No));
    F->cabeca->prox = NULL;
    F->fim = F->cabeca;
}

void enfileirar(Fila *F, int valor) {
    No *novo = (No*) malloc(sizeof(No));
    novo->dado = valor;
    novo->prox = NULL;
    F->fim->prox = novo;
    F->fim = novo;
}

int desenfileirar(Fila *F) {
    if (F->cabeca->prox == NULL) {
        printf("Fila vazia!\n");
        return -1;
    }
    No *removido = F->cabeca->prox;
    int valor = removido->dado;
    F->cabeca->prox = removido->prox;
    if (F->cabeca->prox == NULL) F->fim = F->cabeca;
    free(removido);
    return valor;
}

void mostrar(Fila *F) {
    No *aux = F->cabeca->prox;
    printf("Frente -> ");
    while (aux != NULL) {
        printf("%d ", aux->dado);
        aux = aux->prox;
    }
    printf("<- Fim\n");
}

int main() {
    Fila F;
    inicializar(&F);
    enfileirar(&F, 10);
    enfileirar(&F, 20);
    enfileirar(&F, 30);
    mostrar(&F);
    desenfileirar(&F);
    mostrar(&F);
    return 0;
}

File: AEDS2/estruturas_Abstratas_TAD/fila/filaCabeca.java
package AEDS.AEDS2.estruturas_Abstratas_TAD.fila;

class No {
    int dado;
    No prox;
    No(int d) { dado = d; prox = null; }
}

public class filaCabeca {
    No cabeca;
    No fim;

    filaCabeca() {
        cabeca = new No(0); // nó cabeça
        fim = cabeca;
    }

    void enfileirar(int valor) {
        No novo = new No(valor);
        fim.prox = novo;
        fim = novo;
    }

    int desenfileirar() {
        if (cabeca.prox == null) {
            System.out.println("Fila vazia!");
            return -1;
        }
        No removido = cabeca.prox;
        cabeca.prox = removido.prox;
        if (cabeca.prox == null) fim = cabeca;
        return removido.dado;
    }

    void mostrar() {
        No aux = cabeca.prox;
        System.out.print("Frente -> ");
        while (aux != null) {
            System.out.print(aux.dado + " ");
            aux = aux.prox;
        }
        System.out.println("<- Fim");
    }

    public static void main(String[] args) {
        filaCabeca f = new filaCabeca();
        f.enfileirar(10);
        f.enfileirar(20);
        f.enfileirar(30);
        f.mostrar();
        f.desenfileirar();
        f.mostrar();
    }
}


    
